ark/jsonschema/object.ts

```diff
@@ -0,0 +1,336 @@
+			patternProperties.forEach(([pattern, parsedJsonSchema]) => {
+				if (pattern.test(dataKey) && !parsedJsonSchema.allows(dataValue)) {
+					errors.push(
+						ctx.reject({
+							actual: dataValue,
+							expected: `${parsedJsonSchema.description} as property ${dataKey} matches patternProperty ${pattern}`
+						})
+					)
+				}
+			})
+		})
+		return errors.length === 0
+	}
+}
+
+const parsePropertyNames = (
+	jsonSchema: JsonSchema.ObjectSchema,
+	ctx: TraversalContext
+) => {
+	if (!("propertyNames" in jsonSchema)) return
+
+	const propertyNamesValidator = innerParseJsonSchema.assert(
+		jsonSchema.propertyNames
+	)
+
+	if (
+		"domain" in propertyNamesValidator.json &&
+		propertyNamesValidator.json.domain !== "string"
+	) {
+		ctx.reject({
+			path: ["propertyNames"],
+			actual: `a schema for validating a ${propertyNamesValidator.json.domain as string}`,
+			expected: "a schema for validating a string"
+		})
+	}
+
+	return (data: object, ctx: TraversalContext) => {
+		const errors: false[] = []
+
+		Object.keys(data).forEach(key => {
+			if (!propertyNamesValidator.allows(key)) {
+				errors.push(
+					ctx.reject({
+						message: `property ${key} doesn't adhere to the propertyNames schema of ${propertyNamesValidator.description}`
+					})
+				)
+			}
+		})
+		return errors.length === 0
+	}
+}
+
+const parseRequiredAndOptionalKeys = (
+	jsonSchema: JsonSchema.ObjectSchema,
+	ctx: TraversalContext
+) => {
+	const optionalKeys: string[] = []
+	const requiredKeys: string[] = []
+	if ("properties" in jsonSchema) {
+		if ("required" in jsonSchema) {
+			if (jsonSchema.required.length !== new Set(jsonSchema.required).size) {
+				ctx.reject({
+					expected: "an array of unique strings",
+					path: ["required"]
+				})
+			}
+
+			for (const key of jsonSchema.required) {
+				if (key in jsonSchema.properties) requiredKeys.push(key)
+				else {
+					ctx.reject({
+						actual: key,
+						expected: "a key in the 'properties' object",
+						path: ["required"]
+					})
+				}
+			}
+			for (const key in jsonSchema.properties)
+				if (!jsonSchema.required.includes(key)) optionalKeys.push(key)
+		} else {
+			// If 'required' is not present, all keys are optional
+			optionalKeys.push(...Object.keys(jsonSchema.properties))
+		}
+	} else if ("required" in jsonSchema) {
+		ctx.reject({
+			actual:
+				"an object JSON Schema with 'required' array but no 'properties' object",
+			expected: "a valid object JSON Schema"
+		})
+	}
+
+	return {
+		optionalKeys: optionalKeys.map(key => ({
+			key,
+			value: innerParseJsonSchema.assert(jsonSchema.properties![key]).internal
+		})),
+		requiredKeys: requiredKeys.map(key => ({
+			key,
+			value: innerParseJsonSchema.assert(jsonSchema.properties![key]).internal
+		}))
+	}
+}
+
+const parseAdditionalProperties = (jsonSchema: JsonSchema.ObjectSchema) => {
+	if (!("additionalProperties" in jsonSchema)) return
+
+	const properties = Object.keys(jsonSchema.properties ?? {})
```

Comment:
This would probably be clearer (and very slightly more performant) as something like:

```ts
const properties = jsonSchema.properties ? Object.keys(jsonSchema.properties) : []
```

Reply:
dc43e17a